### PR TITLE
NIFI-12278 Add CodeQL to Static Analysis Job

### PR DESCRIPTION
**Pull request by exceptionfactory**
# Summary

[NIFI-12278](https://issues.apache.org/jira/browse/NIFI-12278) Adds GitHub [CodeQL](https://docs.github.com/en/code-security/code-scanning/introduction-to-code-scanning/about-code-scanning-with-codeql) security scanning to the Static Analysis job in the standard build workflow.

The initial configuration uses the [default](https://docs.github.com/en/code-security/code-scanning/managing-your-code-scanning-configuration/built-in-codeql-query-suites) query suite which limits the scope to avoid false positives.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: .github/workflows/ci-workflow.yml [hunk 1]
```diff
@@ -79,11 +79,12 @@ concurrency:
   cancel-in-progress: true
 
 permissions:
+  security-events: write
```
Comment (ChrisSamo632) [inline .github/workflows/ci-workflow.yml:82]: Be interested to see if you get this to work, I tried to improve the test result reporting, but ran into apparent write permission issues, so raised [NIFI-12192](https://issues.apache.org/jira/browse/NIFI-12192)

Might not be a permissions issue though or could be a different reason that this appears to have failed so far (I didn't spend very long looking into the issue previously, and it was using a different action)

Comment (ChrisSamo632) [inline .github/workflows/ci-workflow.yml:82]: Be interested to see if you get this to work, I tried to improve the test result reporting, but ran into apparent write permission issues, so raised [NIFI-12192](https://issues.apache.org/jira/browse/NIFI-12192)

Might not be a permissions issue though or could be a different reason that this appears to have failed so far (I didn't spend very long looking into the issue previously, and it was using a different action)

**Comment by exceptionfactory** [comment]
Thanks for taking a look at this @ChrisSamo632.

The `security-events: write` permission is required to update the Security section of the repository with findings from CodeQL. The Static Analysis Job completed and published several findings visible to project maintainers, so appears to be working as expected.

**Comment by ChrisSamo632** [comment]
@exceptionfactory yeah, looks like it worked for CodeQL, which is good news. I think it was a different `write` permission I was trying to use with the other action, but it didn't seem to like it, so good that this isn't affecting your PR 🙂

**Comment by joewitt** [comment]
this is slick and will be quite valuable

**Comment by ChrisSamo632** [comment]
@exceptionfactory presumably the plan will be to take this as a baseline, then start raising new tickets to address the findings (particularly the higher priority ones)? Or do you think addressing the identified Critical issues is worth fixing here?

@joewitt absolutely. There are probably other actions we could consider, but obviously we need to be conscious of the extra resources needed to run additional actions (e.g. vulnerability checks, linting) - all as separate tickets/future considerations

**Comment by exceptionfactory** [comment]
@ChrisSamo632 Yes, the goal of this Jira issue and pull request was to introduce the checking, and then address findings in subsequent issues.

This does lengthen the amount of time to run the Static Analysis job, but it still begins with the Checkstyle and Rat validation, and will fail fast if it encounters issues there.

**Review by ChrisSamo632 — APPROVED** [review]
LGTM 👍 will merge, thanks for the further CI updates @exceptionfactory

**Pull request closed without merging**